connectors/sources/gcs.py

```diff
@@ -0,0 +1,311 @@
+
+        Args:
+            resource (aiogoogle.resource.Resource): Resource name for which api call will be made.
+            method (aiogoogle.resource.Method): Method available for the resource.
+            sub_method (aiogoogle.resource.Method, optional): Sub-method available for the method. Defaults to None.
+            full_response (bool, optional): Specifies whether the response is paginated or not. Defaults to False.
+
+        Raises:
+            exception: A instance of an exception class.
+
+        Yields:
+            Dictionary: Response returned by the resource method.
+        """
+        retry_counter = 0
+        while True:
+            try:
+                async with Aiogoogle(
+                    service_account_creds=self.service_account_credentials
+                ) as google_client:
+                    storage_client = await google_client.discover(
+                        api_name=API_NAME, api_version=API_VERSION
+                    )
+                    if not sub_method and STORAGE_EMULATOR_HOST:
+                        # Redirecting calls to fake-gcs-server for e2e test.
+                        storage_client.discovery_document["rootUrl"] = (
+                            STORAGE_EMULATOR_HOST + "/"
+                        )
+                    resource_object = getattr(storage_client, resource)
+                    method_object = getattr(resource_object, method)
+                    if full_response:
+                        first_page_with_next_attached = (
+                            await google_client.as_service_account(
+                                method_object(**kwargs),
+                                full_res=True,
+                            )
+                        )
+                        async for page_items in first_page_with_next_attached:
+                            yield page_items
+                    else:
+                        if sub_method:
+                            method_object = getattr(method_object, sub_method)
+                        yield await google_client.as_service_account(
+                            method_object(**kwargs)
+                        )
+                    break
+            except AttributeError as error:
+                logger.error(
+                    f"Error occurred while generating the resource/method object for an API call. Error: {error}"
+                )
+                raise
+            except Exception as exception:
+                retry_counter += 1
+                if retry_counter > self.retry_count:
+                    raise exception
+                logger.warning(
+                    f"Retry count: {retry_counter} out of {self.retry_count}. Exception: {exception}"
+                )
+                await asyncio.sleep(DEFAULT_WAIT_MULTIPLIER**retry_counter)
+
+    async def ping(self):
+        """Verify the connection with Google Cloud Storage"""
+        try:
+            await self._api_call(
```

Comment:
you can use `anext()` see https://docs.python.org/3/library/functions.html#anext